LSPS1/extensions/jit_channels/LSPS1.jit_channels.yaml

```diff
@@ -0,0 +1,115 @@
+openapi: 3.0.3
+info:
+  description: |
+    API documentation for channel opening request from a Lightning Network Service Provider (LSP). JIT_CHANNELS.
+    
+  version: "1"
+  title: "LSP Spec 1"
+tags:
+- name: "JIT"
+  description: "Extension: `jit_channels`."
+
+
+
+paths:
+
+  /lsp/channels/{id}/jit:
+    get:
+      tags:
+      - JIT
+      summary: "Get information about registered JIT node."
+      description: |
+        Get information about registered JIT node. Feature: `jit_channels`.
+      operationId: "get-channel-jit"
+
+      parameters:
+        - in: path
+          name: id
+          description: "Order id"
+          required: true
+          schema:
+              type: string
+      responses:
+        "200":
+          description: "Channel information"
+          content:
+              application/json:
+                schema:
+                  $ref: "#/components/schemas/ChannelJITResponse"
+        "404":
+          description: "Not found"
+    post:
+      tags:
+      - JIT
+      summary: "Register the JIT node."
+      description: |
+        Register JIT node. Feature: `jit_channels`.
+      operationId: "post-channel-jit"
+
+      parameters:
+        - in: path
+          name: id
+          description: "Order id"
+          required: true
+          schema:
+              type: string
+      requestBody:
+          description: "Information needed to connect to the node."
+          required: true
+          content:
+              application/json:
+                  schema:
+                      $ref: '#/components/schemas/ChannelJITRequest'
+      responses:
+        "200":
+          description: "Channel information"
+          content:
+              application/json:
+                schema:
+                  $ref: "#/components/schemas/ChannelJITResponse"
+        "404":
+          description: "Not found"
+                  
+components:
+    schemas:
+      ChannelJITRequest:
+        type: "object"
+        required:
+          - node_connection_string
+        properties:
+          node_connection_string:
+            description: "The node connection string that the channel will be opened to in case of an incoming payment."
+            example: "03864ef025fde8fb587d989186ce6a4a186895ee44a926bfc370e2c366597a3f8f@3.33.236.230:9735"
```

Comment:
One thing about JIT channels is for mobile nodes is the IP address changes often. The client would then have to connect its lightning node to the LSP in order for the LSP to open a channel to its pubkey. I can't think of a way to describe that other than the readme. Maybe readme is fine.

Reply:
You need to wake up the node anyway for the channel open, don't you? So connecting is not an issue. Readme should be enough. Should we transform this property from `node_connection_string` to `node_connection_string_or_pubkey`?

Reply:
Might also want to be explicit about exactly what the format is for `lsp_connection_string` and `node_connection_string`.  For example, Core Lightning supports the grammar `<node> '@' '[' <ip> ']' ':' <port>`, e.g. `023456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef012@[::1]:42`.  The reason for this is that colons are used in IPv6 addresses, and Core Lightning supports leaving out the port number, i.e. `<node> '@' <ip>` since the BOLT spec specifies you should always use 9735 anyway.

```markdown
The `lsp_connection_string` MUST match one of the following grammars:

* `<node> '@' <ip>`, without any `:` characters inside `<ip>`; this implies the default port for the network, 9735 for Bitcoin Lightning, 19735 for Testnet Bitcoin Lightning.
* `<node> '@' <ip> `:` <port>`, with the last `:` after `@` considered a separator for the port number.
* `<node> '@' '[' <ip> ']'`, with any `:` inside `<ip>` considered part of the IP address, and using the default port for the network; this is needed for IPv6 addresses.
* `<node> '@' '[' <ip> ']' ':' <port>`, with any `:` inside `<ip>` considered part of the IP address.
```

Reply:
I will turn this into a general node connection string definition in the spec (for `lsp_connection_string` and `user_connection_string`).

I would like to remove the lines `<node> '@' <ip>` and `<node> '@' '[' <ip> ']'` that implicitly assume a port. This would simplify the spec and make the connection string more explicit.

I would also like to add support for TOR v3 addresses:
* `<node> '@' tor_onion_v3_address ':' <port>`, with the last `:` after `@` considered a separator for the port number.

> Rationale for only supporting v3: v2 addresses are [deprecated](https://support.torproject.org/onionservices/v2-deprecation/) since July 2021.

Reply:
The user should be allowed to provide a connection_string OR a pubkey. Mobile wallets usually don't have a public IP so they are not able to provide a full connection_string anyway.

I want to change the spec so the peer connection MUST be established by the user and remove the LSP connection requirement completely. My gut feeling tells me that we should add "LSP MAY open a peer connection" though.

Reply:
>Well, they still might try, and this connection string might be bogus. So it should be clear from the spec that these are really two entirely separate paths and you MUST NOT provide a full bogus connection string.

I am not sure I understand why separate paths? In theory, if the user always establishes the peer connection the user_pubkey is enough. The user_connection_string is only necessary in case the LSP wants to establish the connection. Maybe we should fully drop the user_connection_string and only use a user_pubkey instead?

> Yes, I think this makes a lot of sense at least for any NATed case. For any other case it of course might result in some unnecessary delay, as the client can only establish the connection after getting the response, while the LSP could already establish the connection in parallel to sending the response. It's probably negligible though?

In theory, the user can establish a peer connection before they attempt the payment. This way, the channel can be opened instantly as soon as the payment arrived. This should be even faster than establishing the peer connection after the payment arrived.